pinconfigs/gd32f350xx.yml

```diff
@@ -0,0 +1,456 @@
+# GD32F350XX pin definitions
+#
+# Sources:
+# - GD32F350XX Datasheet (Revision 1.5)
+#
+# Pin codes:
+#
+# - 28 pins: G
+# - 32 pins: K
+# - 48 pins: C
+# - 64 pins: R
+#
+# Memory codes:
+#
+# - 16Kb  Flash, 4Kb  SRAM: 4
+# - 32Kb  Flash, 6Kb  SRAM: 6
+# - 64Kb  Flash, 8Kb  SRAM: 8
+# - 64Kb  Flash, 16Kb SRAM: 8(GD32F350R8 only)
+# - 128Kb Flash, 16Kb SRAM: B
+#
+#  PA0 AF1 USART0_CTS, USART1_CTS,
+# Copyright (c) 2021 BrainCo Inc.
+# SPDX-License-Identifier: Apache 2.0
+
+model: af
+
+series: gd32f350
+
+variants:
+  - pincode: G
+    memories: [4]
+  - pincode: G
+    memories: [6]
+  - pincode: G
+    memories: [8]
+  - pincode: K
+    memories: [4]
+  - pincode: K
+    memories: [6]
+  - pincode: K
+    memories: [8]
+  - pincode: C
+    memories: [4]
+  - pincode: C
+    memories: [6]
+  - pincode: C
+    memories: [8, B]
+  - pincode: R
+    memories: [4]
+  - pincode: R
+    memories: [6]
+  - pincode: R
+    memories: [8, B]
+
+pins:
+  PA0:
+    pincodes: [G, K, C, R]
+    include-memories:
+      USART0_CTS: [4]
+      USART1_CTS: [6, 8, B]
+      I2C1_SCL: [8, B]
```

Comment:
have you looked at the optional `signal-configs` section? It's explained here https://github.com/zephyrproject-rtos/hal_gigadevice/tree/main/pinconfigs#af-model. You can specify if a signal has to be excluded for certain memories. I think the proposed approach introduces too much overhead.

Reply:
'exclude-memories' in signal-config is global setting. It worked for the memory doesn't have this peripheral(signals)

From [GD32 MCU Selection Guide](http://gd32mcu.com/download/down/document_id/206/path_type/1), we can find that gd32f350x4 only have usart0, gd32f350B/8/6 have usart0 and usart1.

From below table(copy from gd32f350 datasheet), we can find PA[0-4] AF1 will mapping to USART1 for gd32f350xB/8/6, and USART0 for  gd32f350x4. 
we can use `exclude-memories: [4]` for USART1_, it doesn't have usart1.
we can **NOT** use `exclude-memories[B, 8, 6]` for USART0_, because it have USART0, just mapping in another position.

| | AF1 |
|---  | ----------- |
| PA0 |USART0_CTS (1)  USART1_CTS (2) |
| PA1 |USART0_RTS (1)  USART1_RTS (2) |
| PA2 |USART0_TX (1)    USART1_TX (2) |
| PA3 |USART0_RX (1)    USART1_RX (2) |
| PA4 |USART0_CK (1)    USART1_CK (2) |

Reply:
I think it is better to generate a few extra non-available configs than increasing complexity. A comment on the pinconfig file mentioning this issue can be added to clarify. In the future, we can grab all exceptions and come up with a way to handle them (I'm sure there'll be a few more). Maybe a `exceptions` section where certain rules can be specified to exclude/include signals in a more flexible manner.

Reply:
Can we change the gd32vf350's exclude-memories to use this pattern. Then, we can say, by default, signal available for every memories, but if there have an 'include-memories' in pin config, restriction will applied.

I write an simple tool [gd32-pinctrl-config](https://github.com/BrainCoTech/gd32-pinctrl-config) to generate the header, may help others to generate the pinconfig.

Reply:
I prefer to wait to have a list of all exceptions to decide on what to do. What will happen really is that generated headers will have a few invalid entries, but all possible ones will still be there, so it's not a big issue after all. I'd focus on getting hardware in, even if there are known small issues.

Reply:
I agreed with @gmarull ,

It is better have more cases to get big picture and later improve things. There are other platforms that don't have all pins available and we never get a problem about that. In general, people know HW they are using.

Reply:
OK, add an individual `gd32f350-config` to handle it. remove the `include-meories` from pins.